django/test/client.py

```diff
@@ -55,47 +55,57 @@ def __init__(self, message, last_response):
         self.redirect_chain = last_response.redirect_chain
 
 
-class FakePayload:
+class FakePayload(IOBase):
     """
     A wrapper around BytesIO that restricts what can be read since data from
     the network can't be sought and cannot be read outside of its content
     length. This makes sure that views can't do anything under the test client
     that wouldn't work in real life.
     """
 
-    def __init__(self, content=None):
+    def __init__(self, initial_bytes=None):
         self.__content = BytesIO()
         self.__len = 0
         self.read_started = False
-        if content is not None:
-            self.write(content)
+        if initial_bytes is not None:
+            self.write(initial_bytes)
 
     def __len__(self):
         return self.__len
 
-    def read(self, num_bytes=None):
+    def read(self, size=-1, /):
         if not self.read_started:
             self.__content.seek(0)
             self.read_started = True
-        if num_bytes is None:
-            num_bytes = self.__len or 0
+        if size == -1 or size is None:
+            size = self.__len
         assert (
-            self.__len >= num_bytes
+            self.__len >= size
         ), "Cannot read more than the available bytes from the HTTP incoming data."
-        content = self.__content.read(num_bytes)
-        self.__len -= num_bytes
+        content = self.__content.read(size)
+        self.__len -= len(content)
         return content
 
-    def write(self, content):
+    def readline(self, size=-1, /):
```

Comment:
Not sure what to do about this. I guess we can add a test. We need `.readline()` to exist as we have to cache that in `LimitedStream.__init__()`, but we're not using it anywhere in the test suite at present. The problem is that we cannot simply call `super().readline()` as we need the special handling to ensure we can't write-after-read, and we need to check the size bit before the `assert`.

We hardly use `.readline()` anywhere, but it became necessary that all the provided stream support it because we now directly call it on the stream, rather than wrap a pre-read (via `.read()`) byte string in `BytesIO` and call `.readline()` on that.

I've added some tests to make sure that this works as expected.